server/src/main/java/org/elasticsearch/health/node/tracker/HealthTracker.java

```diff
@@ -8,98 +8,68 @@
- * reported value and can retrieve the current health status when requested.
+ * reported health and can retrieve the current health status when requested.
  *
  * @param <T> the type of the health check result they track
  */
 public abstract class HealthTracker<T> {
-    private static final Logger logger = LogManager.getLogger(HealthTracker.class);
 
-    private final AtomicReference<T> lastReportedValue = new AtomicReference<>();
+    /**
+     * We can "simply" use a volatile field here, as we've ensured only one monitoring instance/thread at a time can update this value.
+     */
+    private volatile T lastReportedHealth;
 
     /**
      * Determine the health info for this health check.
      *
      * @return the health info.
      */
-    public abstract T checkCurrentHealth();
+    protected abstract T determineCurrentHealth();
 
     /**
      * Add the health info to the request builder.
      *
      * @param builder the builder to add the health info to.
      * @param healthInfo the health info to add.
      */
-    public abstract void addToRequestBuilder(UpdateHealthInfoCacheAction.Request.Builder builder, T healthInfo);
+    protected abstract void addToRequestBuilder(UpdateHealthInfoCacheAction.Request.Builder builder, T healthInfo);
 
     /**
-     * Create a new {@link HealthProgress} instance by getting the current last reported value and determining the health info at this time.
-     *
-     * @return the new {@link HealthProgress} instance.
+     * Add the last reported health to the request builder.
      */
-    public HealthProgress<T> trackHealth() {
-        return new HealthProgress<>(this, lastReportedValue.get(), checkCurrentHealth());
+    public void addToRequestBuilder(UpdateHealthInfoCacheAction.Request.Builder builder) {
+        addToRequestBuilder(builder, lastReportedHealth);
     }
 
     /**
-     * Update the last reported health info to <code>current</code>, but only when the value inside <code>lastReportedValue</code>
-     * is equal to <code>previous</code>.
+     * Determine the current health info for this tracker and check if it has changed from the last reported value. When the health has
+     * changed, we'll store the new health as the last reported value.
      *
-     * @param previous the previous value that should be in <code>lastReportedValue</code> at the time of execution.
-     * @param current the value that should be stored in <code>lastReportedValue</code>.
+     * @return whether the health has changed.
      */
-    public void updateLastReportedHealth(T previous, T current) {
-        if (lastReportedValue.compareAndSet(previous, current)) {
-            logger.debug("Health info [{}] successfully sent, last reported value: {}.", current, previous);
+    public boolean trackHealth() {
+        var health = determineCurrentHealth();
+        if (health.equals(lastReportedHealth)) {
```

Comment:
Should we add an assertion that `determinCurrentHealth` does not return `null` ?
i.e. `asssert health != null : "health trackers must return unknown health instead of null"`

Reply:
I like asserts as they _code_ an implicit convention (i.e. make it explicit so aids reading and helps the code reader to not have to go into the contract of `determinCurrentHealth` to see if it can return `null`) and are easy to reason about when they do trip in a test (as the message makes it clear why they tripped)

This is just a personal preference, so feel free to ignore :)

Reply:
I've added the assertion. I'll let the idea/concept of assertions sink in for a bit to determine my long term position on them 😉.